src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4117,220 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
 #endif // DEBUG
 }
 
+//****************************************************************************
+// emitRemoveJumpToNextInst:  Checks all jumps in the jump list to see if they are
+//   unconditional jumps generated at the end of BB_ALWAYS basic blocks and whether
+//   they are now a jump to the next instruction, this can happen when the following
+//   basic block emitted no instructions
+//
+// Returns:
+//   true if any jumps have been removed, otherwise returns false
+//
+// Assumptions:
+//    the jump list must be ordered by increasing igNum+insNo
+//
+bool emitter::emitRemoveJumpToNextInst()
+{
+#ifdef TARGET_XARCH
+    JITDUMP("*************** In emitRemoveJumpToNextInst()\n");
+#ifdef DEBUG
+    if (EMIT_INSTLIST_VERBOSE)
+    {
+        JITDUMP("\nInstruction group list before unconditional jump to next instruction removal:\n\n");
+        emitDispIGlist(true);
+    }
+    if (EMITVERBOSE)
+    {
+        emitDispJumpList();
+    }
+#endif
+
+    int            totalRemovedCount      = 0;
+    UNATIVE_OFFSET totalRemovedSize       = 0;
+    instrDescJmp*  jmp                    = emitJumpList;
+    instrDescJmp*  previousJmp            = nullptr;
+    insGroup*      previousTruncatedGroup = nullptr;
+#if DEBUG
+    UNATIVE_OFFSET previousJumpIgNum  = (UNATIVE_OFFSET)-1;
+    unsigned int   previousJumpInsNum = -1;
+#endif
+
+    while (jmp)
+    {
+        // if the jump is unconditional then it is a candidate
+        if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && jmp->idjIsJmpAlways)
+        {
+            insGroup* jmpGroup = jmp->idjIG;
+#if DEBUG
+            assert((jmpGroup->igFlags & IGF_HAS_ALIGN) == 0);
+            assert(jmpGroup->igNum > previousJumpIgNum || previousJumpIgNum == (UNATIVE_OFFSET)-1 ||
+                   (jmpGroup->igNum == previousJumpIgNum && jmp->idDebugOnlyInfo()->idNum > previousJumpInsNum));
+            previousJumpIgNum  = jmpGroup->igNum;
+            previousJumpInsNum = jmp->idDebugOnlyInfo()->idNum;
+#endif
+            // target group is not bound yet so use the cookie to fetch it
+            insGroup* targetGroup = (insGroup*)emitCodeGetCookie(jmp->idAddr()->iiaBBlabel);
+
+            if (targetGroup != nullptr && jmpGroup->igNext == targetGroup)
+            {
+                unsigned instructionCount = jmpGroup->igInsCnt;
+
+                if (instructionCount > 1)
+                {
+                    instrDesc* instructionDescriptor = nullptr;
+
+                    {
+                        BYTE* dataPtr = jmpGroup->igData;
```

Comment:
I am not sure why we need to retrieve `instructionDescriptor` here? If it is `idjIsJmpAlways`, shouldn't it always be the last instruction? Did you try putting the logic of locating the instruction under `#ifdef` and found cases where `jmp != instructionDescriptor`? . If yes, we need to fix those cases and then just rely on  `jmp` to do the operations. `instructionDescriptor` part can stay but under `#ifdef`.

Reply:
Testing showed no errors with single instruction ig's having their jump removed and that there were examples in the tests. Consequently this is now debug only code as requested.

Reply:
Changing this to assert breaks several non-windows builds. I put in a dump temporarily to get information from the build for them. I have no way to work on non-windows builds.

```
/Users/runner/work/1/s/dotnet.sh /Users/runner/work/1/s/artifacts/bin/coreclr/OSX.x64.Debug/crossgen2/crossgen2.dll -o:/Users/runner/work/1/s/artifacts/bin/coreclr/OSX.x64.Debug/System.Private.CoreLib.dll -r:/Users/runner/work/1/s/artifacts/bin/coreclr/OSX.x64.Debug/IL/*.dll --targetarch:x64 -O --verify-type-and-field-layout /Users/runner/work/1/s/artifacts/bin/coreclr/OSX.x64.Debug/IL/System.Private.CoreLib.dll
  about to assert, dumping jump list
  G_M79203_IG05:        ; gcVars= {V00 V02}, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, gcvars, byref
  
  000051 mov      rdi, gword ptr [V00 rbp-40H]
  000055 lea      rax, bword ptr [rdi+120]
  000059 cmp      gword ptr [rax], 0
  00005D je       L_M79203_BB12
  000063 mov      rax, gword ptr [rax]
  000066 cmp      dword ptr [rax+32], 0
  00006A setne    al
  00006D movzx    rax, al
  000070 test     al, al
  000072 je       L_M79203_BB12
  000078 jmp      L_M79203_BB11
  00007D jmp      L_M79203_BB12
  
  /Users/runner/work/1/s/src/coreclr/jit/emit.cpp:4200
  Assertion failed 'instructionDescriptor->idIns() == jmp->idIns() && jmp == instructionDescriptor' in 'ThreadPoolValueTaskSource:ExecuteInternal():this' during 'Generate code' (IL size 511; hash 0xaf9afbb3; FullOpts)
```

Reply:
I added in a check for always jumps in the instruction dump function. Based on the updated dump info above it looks like two BB's got combined and that might be for the same reason you've already got a PR in progress for.

Reply:
Your PR fixes this precise issue. This leave the questions. How confident can we be that the same problem cannot ever occur again? 

I prefer caution. I can see that there is no technical barrier to it happening again via some other route we haven't seen, I also see that we can never have complete coverage of every pattern users are capable of producing. In release mode this can cause existing correct but suboptimal code to be changed into incorrect code I advise leaving the loop in place. I recognise it would be better for performance not to loop over the instructions but it may be incorrect and correct is more important that fast.